Mathlib/Algebra/QuadraticAlgebra.lean

```diff
@@ -0,0 +1,418 @@
+/-
+Copyright (c) 2025 Kenny Lau. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yunzhou Xie, Kenny Lau, Jiayang Hong
+-/
+
+import Mathlib.LinearAlgebra.Dimension.StrongRankCondition
+import Mathlib.LinearAlgebra.FreeModule.Finite.Basic
+
+/-!
+
+# Quadratic Algebra
+
+In this file we define the quadratic algebra `QuadraticAlgebra R a b` over a commutative ring `R`,
+and define some algebraic structures on it.
+
+## Main definitions
+
+* `QuadraticAlgebra R a b`:
+  [Bourbaki, *Algebra I*][bourbaki1989] with coefficients `a`, `b` in `R`.
+
+## Tags
+
+Quadratic algebra, quadratic extension
+
+-/
+
+universe u
+
+/-- Quadratic algebra over a type with fixed coefficient where $i^2 = a + bi$, implemented as
+a structure with two fields, `re` and `im`. When `R` is a commutative ring, this is isomorphic to
+`R[X]/(X^2-b*X-a)`. -/
```

Comment:
Can it not be obtained as a quotient in general? Also, where is that `lift` operation that you get from identifying `QuadraticAlgebra` to the quotient?

Reply:
The reason this is not quotient is the same reason why Zsqrtd and C are not quotients -- i.e. for better (decidable) equality

Reply:
Ah, I misunderstood, you meant this mathematically -- in which case, yes but you will need a relation in general, to be the relation generated by X^2 ~ a+bX

Reply:
Is the ask for `Complex.lift` but for `QuadraticAlgebra`?

Reply:
Then can you define it as a quotient? We've got plenty of API for quotients, and it seems a shame to not use it. Here is an example where @Kiolt, @erdOne and I worked out the ring corresponding to SO(2): https://github.com/YaelDillies/Toric/blob/master/Toric/Examples/SO2.lean#L37

Reply:
The benefit is that you plug into the usual quotient machinery. In general, the "mathematically quotients" which are "Lean quotients" are of two kinds:
1. the fundamental ones, like `Int`, `Rat`, that are used as internals to tactics
2. the historically early ones, like `ZMod` and `Zsqrtd`, which are either a bit weird to use on the edges, or not used at all anymore

I am afraid you are on your way to creating a new instance of the second kind, which would be sad!

I see no issue with creating a `norm_num` tactic for quotients. After all, your `norm_num_quat` isn't going to act on elements of the quadratic algebra, but on *expressions* thereof, and you can normalise your expressions at each step, the same way as you're currently doing here.

Reply:
I don't think this is fair: ZMod is used quite a lot, and Zsqrtd is not used because it's mathematically wrong! The argument "it seems a shame to not use it" doesn't seem like a particularly compelling argument. I'm curious what number theorists think about this: @riccardobrasca, perhaps?

Reply:
Another argument: the two field structure does not require Semiring

Reply:
cf that recent thread expounding why `IsAdjoinRoot` is in fact not such a great idea: https://leanprover.zulipchat.com/#narrow/channel/287929-mathlib4/topic/.60IsAdjoinRoot.60

Reply:
I recently used Zsqrtd in [this PR](https://github.com/leanprover-community/mathlib4/pull/28787/files#diff-28e92735f040065df8555b2dd8ecdd0feb47ec6758d72e878edbe7283420b0fcR17) and found it super convenient to use, and I'm in favor of the current state of this PR defining QuadraticAlgebra as a generalization of Zsqrtd. You might try to rewrite the example with `AdjoinRoot (X^2-1 : Z[X])` and see for yourself how much longer it becomes.